File: Python/LongestSubstringWithoutRepeatingCharacters.py
'''
Question : Given a string s, find the length of the longest substring without repeating characters.

Link : https://leetcode.com/problems/longest-substring-without-repeating-characters/
'''

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        memo = {}
        start , end , maxlength = 0 , 0 , 1
        
        if len(s) == 0:
            return 0 
        
        while start < len(s):
            
            if s[start] in memo:
                end = max(end , memo[s[start]]+1)
            
            maxlength = max(maxlength , start-end+1)
            memo[s[start]] = start
            start += 1
            
        return maxlength
                        

File: Java/LongestSubstringWithoutRepeatingCharacters.java
/*
Question : Given a string s, find the length of the longest substring without repeating characters.

Link : https://leetcode.com/problems/longest-substring-without-repeating-characters/
*/

class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int a = 0;
        int b = 0;
        int max = 0;
        
        HashSet <Character> map = new HashSet();
        
        while(b < s.length()) {
            if(!map.contains(s.charAt(b))) {
                map.add(s.charAt(b));
                b++;
                max = Math.max(map.size() , max);
            }     
            else {
                map.remove(s.charAt(a));
                a++;
            }
        }     
        return max;
    }
}
